chirp-impl/src/main/java/sample/chirper/chirp/impl/JpaReadSideImpl.java

```diff
@@ -0,0 +1,124 @@
+                PMap<Class<? extends Event>, BiConsumer<EntityManager, ? extends Event>> eventHandlers) {
+            this.readSideId = readSideId;
+            this.eventHandlers = eventHandlers;
+        }
+
+        @Override
+        public CompletionStage<Offset> prepare(AggregateEventTag<Event> tag) {
+            this.tag = tag;
+            return FutureConverters.toJava(offsetStore.prepare(readSideId, tag.tag()))
+                    .thenApply(offsetDao -> {
+                        this.offsetDao = offsetDao;
+                        return offsetDao.loadedOffset();
+                    });
+        }
+
+        @Override
+        public Flow<Pair<Event, Offset>, Done, ?> handle() {
+            return Flow.<Pair<Event, Offset>>create().mapAsync(1, eventAndOffset -> {
+                Event event = eventAndOffset.first();
+                Offset offset = eventAndOffset.second();
+                @SuppressWarnings("unchecked") Class<Event> eventClass = (Class<Event>) event.getClass();
+                @SuppressWarnings("unchecked") BiConsumer<EntityManager, Event> eventHandler =
+                        (BiConsumer<EntityManager, Event>) eventHandlers.get(eventClass);
+                if (eventHandler != null) {
+                    return jpa.withTransaction(entityManager -> {
+                        eventHandler.accept(entityManager, event);
+                        updateOffset(entityManager, offset);
+                        return Done.getInstance();
+                    });
+                } else {
+                    /*
+                        if (log.isDebugEnabled)
+                          log.debug("Unhandled event [{}]", eventClass.getName())
+                     */
+                    return CompletableFuture.completedFuture(Done.getInstance());
+                }
+            });
+        }
+
+        private void updateOffset(EntityManager entityManager, Offset offset) {
+            Long sequenceOffset = offset instanceof Offset.Sequence ?
+                    ((Offset.Sequence) offset).value() : null;
+            String timeUuidOffset = offset instanceof Offset.TimeBasedUUID ?
+                    ((Offset.TimeBasedUUID) offset).value().toString() : null;
+            Iterable<String> sqlStatements =
+                    asJavaIterable(offsetDao.updateOffsetQuery(offset).statements());
+
+            for (String statement : sqlStatements) {
+                System.err.println("Executing: " + statement);
+                // NOTE: The order of parameters here depends on the order chosen
+                // by Slick, based on the table definition in JdbcOffsetStore.
+                entityManager.createNativeQuery(statement)
+                        .setParameter(1, readSideId)
+                        .setParameter(2, tag.tag())
+                        .setParameter(3, sequenceOffset)
+                        .setParameter(4, timeUuidOffset)
+                        .executeUpdate();
+            }
```

Comment:
"The problem with using db.run is that it will then claim a second connection and not share the transaction." Oh, you're right. We're restricted to running sql code using `JPA` only.